snuba/query/allocation_policies/bytes_scanned_rejecting_policy.py

```diff
@@ -0,0 +1,259 @@
+        if referrer in _PASS_THROUGH_REFERRERS:
+            return QuotaAllowance(True, self.max_threads, {})
+        scan_limit = self.__get_scan_limit(
+            customer_tenant_key, customer_tenant_value, referrer
+        )
+
+        timestamp, granted_quotas = _RATE_LIMITER.check_within_quotas(
+            [
+                RequestedQuota(
+                    self.runtime_config_prefix,
+                    # request a big number because we don't know how much we actually
+                    # will use in this query. this doesn't use up any quota, we just want to know how much is left
+                    UNREASONABLY_LARGE_NUMBER_OF_BYTES_SCANNED_PER_QUERY,
+                    [
+                        Quota(
+                            # TODO: Make window configurable but I don't know exactly how the rate limiter
+                            # reacts to such changes
+                            window_seconds=self.WINDOW_SECONDS,
+                            granularity_seconds=self.WINDOW_GRANULARITY_SECONDS,
+                            limit=scan_limit,
+                            prefix_override=f"{self.runtime_config_prefix}-{customer_tenant_key}-{customer_tenant_value}-{referrer}",
+                        )
+                    ],
+                ),
+            ]
+        )
+        explanation: dict[str, Any] = {}
+        granted_quota = granted_quotas[0]
+        if granted_quota.granted <= 0:
+            explanation[
+                "reason"
+            ] = f"{customer_tenant_key} {customer_tenant_value} is over the bytes scanned limit of {scan_limit} for referrer {referrer}"
+            explanation["granted_quota"] = granted_quota.granted
+            explanation["limit"] = scan_limit
+            return QuotaAllowance(False, self.max_threads, explanation)
+        return QuotaAllowance(True, self.max_threads, {"reason": "within_limit"})
+
+    def _get_bytes_scanned_in_query(
+        self, tenant_ids: dict[str, str | int], result_or_error: QueryResultOrError
+    ) -> int:
+        progress_bytes_scanned = cast(int, result_or_error.query_result.result.get("profile", {}).get("progress_bytes", None))  # type: ignore
+        if isinstance(progress_bytes_scanned, (int, float)):
+            self.metrics.increment(
+                "progress_bytes_scanned",
+                progress_bytes_scanned,
+                tags={"referrer": str(tenant_ids.get("referrer", "no_referrer"))},
+            )
+
+        return progress_bytes_scanned
+
+    def _update_quota_balance(
+        self,
+        tenant_ids: dict[str, str | int],
+        query_id: str,
+        result_or_error: QueryResultOrError,
+    ) -> None:
+        ids_are_valid, why = self._are_tenant_ids_valid(tenant_ids)
+        if not ids_are_valid:
+            # we already logged the reason before the query
+            return
+        if self.is_cross_org_query(tenant_ids):
+            return
+        if result_or_error.error:
```

Comment:
I'm not sure we should skip the update on an error. I wouldn't want someone to send huge queries that time out, and all those time outs don't count against the limit.

Reply:
That's a good call, how about a configurable timeout query bytes_scanned penalization?